File: LeetCode_290/main.cpp
/**
    LeetCode_290
    https://leetcode.com/problems/word-pattern/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <sstream>

using namespace std;

class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, int> ump;
        unordered_map<string, int> ums;
        stringstream ss(str);
        int n = pattern.size(), i = 0;
        for (string tmp; ss >> tmp; ++i) {
            if (i == n || ump[pattern[i]] != ums[tmp])
                return false;
            ump[pattern[i]] = ums[tmp] = i + 1;
        }
        return i == n;
    }
};

int main() {
    return 0;
}

File: LeetCode_290/main.go
/**
  LeetCode_309
  https://leetcode.com/problems/word-pattern/
  MatrixDeity, 2023.
*/

package leetcode290

import "strings"

func wordPattern(pattern string, s string) bool {
	charsCounter := map[byte]int{}
	wordsCounter := map[string]int{}
	n := len(pattern)
	lastI := 0

	for i, word := range strings.Split(s, " ") {
		if i == n || charsCounter[pattern[i]] != wordsCounter[word] {
			return false
		}
		charsCounter[pattern[i]] = i + 1
		wordsCounter[word] = i + 1
		lastI = i + 1
	}

	return lastI == n
}
